lib_eio/unix/thread_pool.ml

```diff
@@ -0,0 +1,155 @@
+(* This thread pool does not spawn threads in advance,
+   but up to [max_standby_systhreads_per_domain] threads are
+   kept alive to wait for more work to arrive.
+   This number was chosen somewhat arbitrarily but benchmarking
+   shows it to be a good compromise. *)
+let max_standby_systhreads_per_domain = 20
+
+type job =
+| New
+| Exit
+| Job : {
+  fn: unit -> 'a;
+  enqueue: ('a, exn) result -> unit;
+} -> job
+
+(* Mailbox with blocking semaphore *)
+module Mailbox = struct
+  type t = {
+    available: Semaphore.Binary.t;
+    mutable cell: job;
+  }
+
+  let create () = { available = Semaphore.Binary.make false; cell = New }
+
+  let dummy = create ()
+
+  let put mbox x =
+    (* The Semaphore contains an atomic frontier,
+       therefore [cell] does not need to be an atomic *)
+    mbox.cell <- x;
+    Semaphore.Binary.release mbox.available
+
+  let take mbox =
+    Semaphore.Binary.acquire mbox.available;
+    mbox.cell
+end
+
+type t = {
+  threads: Mailbox.t array;
+  available: bool array;
+  mutable num_available: int;
+  mutable terminating: bool;
+}
+
+let create () = {
+    threads = Array.init max_standby_systhreads_per_domain (fun _ -> Mailbox.dummy);
+    available = Array.init max_standby_systhreads_per_domain (fun _ -> false);
+    num_available = 0;
+    terminating = false;
+  }
+
+let terminate pool =
+  pool.terminating <- true;
+  if pool.num_available > 0 then
+  for i = 0 to max_standby_systhreads_per_domain - 1 do
+    if Array.unsafe_get pool.available i
+    then Mailbox.put (Array.unsafe_get pool.threads i) Exit
+  done
+
+(* Can't [assert false] in [@poll error] code because it would allocate *)
+exception Assertion_failure
+
+(* This function is called from within a [@poll error] function,
+   so we cannot allocate, call OCaml functions, or use for/while loops.
+   See https://discuss.ocaml.org/t/using-poll-error-attribute-to-implement-systhread-safe-data-structures/12804
+   The only option left to us is to unroll the loop by hand.
```

Comment:
You could also use lock-free programming techniques instead of unrolling.  The critical part is when you update the state.  At that point you need to atomically check that the state is consistent with the update (`compare`) before the update (`set`) is done.

Essentially something like this:

```ocaml
(** Atomically determines whether to keep a thread. *)
let[@poll never] atomically_should_keep pool =
  pool.num_available < Array.length pool.available &&
  not pool.terminating

(** Non-atomically tries to find a slot for a thread to keep. *)
let try_find_available_slot pool =
  let rec loop i =
    if i < Array.length pool.available then
      if Array.unsafe_get pool.available i then
        Some i
      else
        loop (i + 1)
    else
      None
  in
  loop 0

(** Tries to atomically keep the thread at specified slot. *)
let[@poll error] try_atomically_keep pool i mbox =
  not pool.terminating &&
  Array.unsafe_get pool.available i && begin
    Array.unsafe_set pool.available i true;
    Array.unsafe_set pool.threads i mabox;
    pool.num_available <- pool.num_available + 1;
    true
  end

let rec keep_thread_or_exit pool mbox =
  if atomically_should_keep pool then
    match try_find_available_slot pool with
    | Some i ->
      if not (try_atomically_keep pool i mbox) then begin
        keep_thread_or_exit pool mbox
      end
    | None ->
      keep_thread_or_exit pool mbox
  else
    raise Thread.Exit
```

The above is only to give an idea.  I have neither tested nor compiled the above.

Reply:
That's pretty much what I had in the first version (with atomics)! I like the idea of using it with `[@poll error]` code, that's neat.